File: crates/agentpin/src/resolver.rs
//! Discovery resolver abstraction.
//!
//! Provides pluggable discovery mechanisms beyond the standard
//! `https://{domain}/.well-known/agent-identity.json` endpoint.

use std::collections::HashMap;
use std::path::{Path, PathBuf};

use crate::error::Error;
use crate::types::bundle::TrustBundle;
use crate::types::discovery::DiscoveryDocument;
use crate::types::revocation::RevocationDocument;

// ---------------------------------------------------------------------------
// Sync resolver trait (always available)
// ---------------------------------------------------------------------------

/// Resolve discovery and revocation documents for a given domain.
///
/// Implementations can fetch from `.well-known` URLs, the local filesystem,
/// an in-memory trust bundle, or any other source.
pub trait DiscoveryResolver: Send + Sync {
    /// Return the discovery document for `domain`.
    fn resolve_discovery(&self, domain: &str) -> Result<DiscoveryDocument, Error>;

    /// Return the revocation document for `domain`, if available.
    ///
    /// The default implementation returns `Ok(None)`.
    fn resolve_revocation(
        &self,
        domain: &str,
        _discovery: &DiscoveryDocument,
    ) -> Result<Option<RevocationDocument>, Error> {
        let _ = domain;
        Ok(None)
    }
}

// ---------------------------------------------------------------------------
// Async resolver trait (fetch-gated)
// ---------------------------------------------------------------------------

#[cfg(feature = "fetch")]
/// Async equivalent of [`DiscoveryResolver`].
///
/// Gated behind the `fetch` feature because it brings in `async-trait`.
#[async_trait::async_trait]
pub trait AsyncDiscoveryResolver: Send + Sync {
    async fn resolve_discovery(&self, domain: &str) -> Result<DiscoveryDocument, Error>;

    async fn resolve_revocation(
        &self,
        domain: &str,
        _discovery: &DiscoveryDocument,
    ) -> Result<Option<RevocationDocument>, Error> {
        let _ = domain;
        Ok(None)
    }
}

// ---------------------------------------------------------------------------
// WellKnownResolver (fetch-gated — wraps existing HTTP fetchers)
// ---------------------------------------------------------------------------

#[cfg(feature = "fetch")]
/// Fetches documents from the standard `.well-known` HTTPS endpoint.
pub struct WellKnownResolver;

#[cfg(feature = "fetch")]
#[async_trait::async_trait]
impl AsyncDiscoveryResolver for WellKnownResolver {
    async fn resolve_discovery(&self, domain: &str) -> Result<DiscoveryDocument, Error> {
        crate::discovery::fetch_discovery_document(domain).await
    }

    async fn resolve_revocation(
        &self,
        _domain: &str,
        discovery: &DiscoveryDocument,
    ) -> Result<Option<RevocationDocument>, Error> {
        if let Some(ref endpoint) = discovery.revocation_endpoint {
            let doc = crate::revocation::fetch_revocation_document(endpoint).await?;
            Ok(Some(doc))
        } else {
            Ok(None)
        }
    }
}

// ---------------------------------------------------------------------------
// LocalFileResolver (sync — reads from filesystem)
// ---------------------------------------------------------------------------

/// Reads discovery documents from a local directory.
///
/// Expects files named `{domain}.json` under `discovery_dir`.  Optionally
/// reads revocation documents from `{domain}.revocations.json` in the same
/// directory (or a separate `revocation_dir`).
pub struct LocalFileResolver {
    discovery_dir: PathBuf,
    revocation_dir: Option<PathBuf>,
}

impl LocalFileResolver {
    pub fn new(discovery_dir: &Path, revocation_dir: Option<&Path>) -> Self {
        Self {
            discovery_dir: discovery_dir.to_path_buf(),
            revocation_dir: revocation_dir.map(|p| p.to_path_buf()),
        }
    }
}

impl DiscoveryResolver for LocalFileResolver {
    fn resolve_discovery(&self, domain: &str) -> Result<DiscoveryDocument, Error> {
        let path = self.discovery_dir.join(format!("{}.json", domain));
        let data = std::fs::read_to_string(&path)
            .map_err(|e| Error::Discovery(format!("Cannot read {}: {}", path.display(), e)))?;
        let doc: DiscoveryDocument = serde_json::from_str(&data)?;
        Ok(doc)
    }

    fn resolve_revocation(
        &self,
        domain: &str,
        _discovery: &DiscoveryDocument,
    ) -> Result<Option<RevocationDocument>, Error> {
        let dir = self.revocation_dir.as_ref().unwrap_or(&self.discovery_dir);
        let path = dir.join(format!("{}.revocations.json", domain));
        if !path.exists() {
            return Ok(None);
        }
        let data = std::fs::read_to_string(&path)
            .map_err(|e| Error::Discovery(format!("Cannot read {}: {}", path.display(), e)))?;
        let doc: RevocationDocument = serde_json::from_str(&data)?;
        Ok(Some(doc))
    }
}

// ---------------------------------------------------------------------------
// TrustBundleResolver (sync — in-memory lookup)
// ---------------------------------------------------------------------------

/// Resolves documents from a pre-loaded [`TrustBundle`].
pub struct TrustBundleResolver {
    discovery: HashMap<String, DiscoveryDocument>,
    revocations: HashMap<String, RevocationDocument>,
}

impl TrustBundleResolver {
    /// Build from a [`TrustBundle`].
    pub fn new(bundle: &TrustBundle) -> Self {
        let mut discovery = HashMap::new();
        for doc in &bundle.documents {
            discovery.insert(doc.entity.clone(), doc.clone());
        }
        let mut revocations = HashMap::new();
        for doc in &bundle.revocations {
            revocations.insert(doc.entity.clone(), doc.clone());
        }
        Self {
            discovery,
            revocations,
        }
    }

    /// Build from a JSON string representing a [`TrustBundle`].
    pub fn from_json(json: &str) -> Result<Self, Error> {
        let bundle: TrustBundle = serde_json::from_str(json)?;
        Ok(Self::new(&bundle))
    }
}

impl DiscoveryResolver for TrustBundleResolver {
    fn resolve_discovery(&self, domain: &str) -> Result<DiscoveryDocument, Error> {
        self.discovery
            .get(domain)
            .cloned()
            .ok_or_else(|| Error::Discovery(format!("Domain '{}' not in trust bundle", domain)))
    }

    fn resolve_revocation(
        &self,
        domain: &str,
        _discovery: &DiscoveryDocument,
    ) -> Result<Option<RevocationDocument>, Error> {
        Ok(self.revocations.get(domain).cloned())
    }
}

// ---------------------------------------------------------------------------
// ChainResolver (sync — tries resolvers in order)
// ---------------------------------------------------------------------------

/// Composite resolver that tries a sequence of resolvers in order until one
/// succeeds.
pub struct ChainResolver {
    resolvers: Vec<Box<dyn DiscoveryResolver>>,
}

impl ChainResolver {
    pub fn new(resolvers: Vec<Box<dyn DiscoveryResolver>>) -> Self {
        Self { resolvers }
    }
}

impl DiscoveryResolver for ChainResolver {
    fn resolve_discovery(&self, domain: &str) -> Result<DiscoveryDocument, Error> {
        let mut last_err = Error::Discovery("No resolvers configured".to_string());
        for resolver in &self.resolvers {
            match resolver.resolve_discovery(domain) {
                Ok(doc) => return Ok(doc),
                Err(e) => last_err = e,
            }
        }
        Err(last_err)
    }

    fn resolve_revocation(
        &self,
        domain: &str,
        discovery: &DiscoveryDocument,
    ) -> Result<Option<RevocationDocument>, Error> {
        for resolver in &self.resolvers {
            match resolver.resolve_revocation(domain, discovery) {
                Ok(Some(doc)) => return Ok(Some(doc)),
                Ok(None) => continue,
                Err(_) => continue,
            }
        }
        Ok(None)
    }
}

// ---------------------------------------------------------------------------
// Tests
// ---------------------------------------------------------------------------

#[cfg(test)]
mod tests {
    use super::*;
    use crate::discovery::build_discovery_document;
    use crate::revocation::build_revocation_document;
    use crate::types::discovery::EntityType;

    fn make_discovery(domain: &str) -> DiscoveryDocument {
        build_discovery_document(
            domain,
            EntityType::Maker,
            vec![crate::jwk::Jwk {
                kid: "test-key".to_string(),
                kty: "EC".to_string(),
                crv: "P-256".to_string(),
                x: "x".to_string(),
                y: "y".to_string(),
                use_: "sig".to_string(),
                key_ops: None,
                exp: None,
            }],
            vec![],
            2,
            "2026-01-15T00:00:00Z",
        )
    }

    // -- TrustBundleResolver -------------------------------------------------

    #[test]
    fn test_trust_bundle_resolver_hit() {
        let bundle = TrustBundle {
            agentpin_bundle_version: "0.1".to_string(),
            created_at: "2026-02-10T00:00:00Z".to_string(),
            documents: vec![make_discovery("example.com")],
            revocations: vec![],
        };
        let resolver = TrustBundleResolver::new(&bundle);
        let doc = resolver.resolve_discovery("example.com").unwrap();
        assert_eq!(doc.entity, "example.com");
    }

    #[test]
    fn test_trust_bundle_resolver_miss() {
        let bundle = TrustBundle::new("2026-02-10T00:00:00Z");
        let resolver = TrustBundleResolver::new(&bundle);
        assert!(resolver.resolve_discovery("missing.com").is_err());
    }

    #[test]
    fn test_trust_bundle_resolver_revocation() {
        let rev = build_revocation_document("example.com");
        let bundle = TrustBundle {
            agentpin_bundle_version: "0.1".to_string(),
            created_at: "2026-02-10T00:00:00Z".to_string(),
            documents: vec![make_discovery("example.com")],
            revocations: vec![rev],
        };
        let resolver = TrustBundleResolver::new(&bundle);
        let disc = resolver.resolve_discovery("example.com").unwrap();
        let rev = resolver.resolve_revocation("example.com", &disc).unwrap();
        assert!(rev.is_some());
    }

    #[test]
    fn test_trust_bundle_resolver_from_json() {
        let bundle = TrustBundle {
            agentpin_bundle_version: "0.1".to_string(),
            created_at: "2026-02-10T00:00:00Z".to_string(),
            documents: vec![make_discovery("example.com")],
            revocations: vec![],
        };
        let json = serde_json::to_string(&bundle).unwrap();
        let resolver = TrustBundleResolver::from_json(&json).unwrap();
        assert!(resolver.resolve_discovery("example.com").is_ok());
    }

    // -- LocalFileResolver ---------------------------------------------------

    #[test]
    fn test_local_file_resolver() {
        let dir = tempfile::tempdir().unwrap();
        let doc = make_discovery("local.example.com");
        let path = dir.path().join("local.example.com.json");
        std::fs::write(&path, serde_json::to_string_pretty(&doc).unwrap()).unwrap();

        let resolver = LocalFileResolver::new(dir.path(), None);
        let resolved = resolver.resolve_discovery("local.example.com").unwrap();
        assert_eq!(resolved.entity, "local.example.com");
    }

    #[test]
    fn test_local_file_resolver_missing() {
        let dir = tempfile::tempdir().unwrap();
        let resolver = LocalFileResolver::new(dir.path(), None);
        assert!(resolver.resolve_discovery("missing.com").is_err());
    }

    #[test]
    fn test_local_file_resolver_revocation() {
        let dir = tempfile::tempdir().unwrap();
        let doc = make_discovery("local.example.com");
        let rev = build_revocation_document("local.example.com");
        std::fs::write(
            dir.path().join("local.example.com.json"),
            serde_json::to_string(&doc).unwrap(),
        )
        .unwrap();
        std::fs::write(
            dir.path().join("local.example.com.revocations.json"),
            serde_json::to_string(&rev).unwrap(),
        )
        .unwrap();

        let resolver = LocalFileResolver::new(dir.path(), None);
        let resolved = resolver
            .resolve_revocation("local.example.com", &doc)
            .unwrap();
        assert!(resolved.is_some());
    }

    // -- ChainResolver -------------------------------------------------------

    #[test]
    fn test_chain_resolver_first_wins() {
        let bundle_a = TrustBundle {
            agentpin_bundle_version: "0.1".to_string(),
            created_at: "2026-02-10T00:00:00Z".to_string(),
            documents: vec![make_discovery("a.com")],
            revocations: vec![],
        };
        let bundle_b = TrustBundle {
            agentpin_bundle_version: "0.1".to_string(),
            created_at: "2026-02-10T00:00:00Z".to_string(),
            documents: vec![make_discovery("b.com")],
            revocations: vec![],
        };

        let chain = ChainResolver::new(vec![
            Box::new(TrustBundleResolver::new(&bundle_a)),
            Box::new(TrustBundleResolver::new(&bundle_b)),
        ]);

        assert!(chain.resolve_discovery("a.com").is_ok());
        assert!(chain.resolve_discovery("b.com").is_ok());
        assert!(chain.resolve_discovery("c.com").is_err());
    }

    #[test]
    fn test_chain_resolver_fallthrough() {
        let empty = TrustBundle::new("2026-02-10T00:00:00Z");
        let has_doc = TrustBundle {
            agentpin_bundle_version: "0.1".to_string(),
            created_at: "2026-02-10T00:00:00Z".to_string(),
            documents: vec![make_discovery("example.com")],
            revocations: vec![],
        };

        let chain = ChainResolver::new(vec![
            Box::new(TrustBundleResolver::new(&empty)),
            Box::new(TrustBundleResolver::new(&has_doc)),
        ]);

        let doc = chain.resolve_discovery("example.com").unwrap();
        assert_eq!(doc.entity, "example.com");
    }
}

File: go/pkg/resolver/resolver.go
// Package resolver provides discovery / revocation resolution strategies
// (well-known HTTPS, local file, trust bundle, chain).
package resolver

import (
	"encoding/json"
	"errors"
	"fmt"
	"net/http"
	"os"
	"path/filepath"

	"github.com/ThirdKeyAi/agentpin/go/pkg/discovery"
	"github.com/ThirdKeyAi/agentpin/go/pkg/revocation"
	"github.com/ThirdKeyAi/agentpin/go/pkg/types"
)

// DiscoveryResolver resolves discovery and revocation documents for a
// domain. Implementations may fetch from HTTPS, the local filesystem, a
// pre-loaded trust bundle, or anything else.
type DiscoveryResolver interface {
	ResolveDiscovery(domain string) (*types.DiscoveryDocument, error)
	ResolveRevocation(domain string, discovery *types.DiscoveryDocument) (*types.RevocationDocument, error)
}

// WellKnownResolver fetches documents from the standard `.well-known` HTTPS
// endpoint.
type WellKnownResolver struct {
	Client *http.Client
}

// NewWellKnownResolver returns a resolver using the default HTTP client.
func NewWellKnownResolver() *WellKnownResolver { return &WellKnownResolver{} }

// ResolveDiscovery implements DiscoveryResolver.
func (r *WellKnownResolver) ResolveDiscovery(domain string) (*types.DiscoveryDocument, error) {
	return discovery.FetchDiscoveryDocument(r.Client, domain)
}

// ResolveRevocation implements DiscoveryResolver.
func (r *WellKnownResolver) ResolveRevocation(_ string, doc *types.DiscoveryDocument) (*types.RevocationDocument, error) {
	if doc == nil || doc.RevocationEndpoint == "" {
		return nil, nil
	}
	return revocation.FetchRevocationDocument(r.Client, doc.RevocationEndpoint)
}

// LocalFileResolver reads discovery documents from a local directory in
// `{domain}.json` form, and (optionally) revocation documents from
// `{domain}.revocations.json`.
type LocalFileResolver struct {
	DiscoveryDir  string
	RevocationDir string
}

// NewLocalFileResolver returns a resolver rooted at discoveryDir. If
// revocationDir is empty, revocations are read from discoveryDir.
func NewLocalFileResolver(discoveryDir, revocationDir string) *LocalFileResolver {
	return &LocalFileResolver{DiscoveryDir: discoveryDir, RevocationDir: revocationDir}
}

// ResolveDiscovery implements DiscoveryResolver.
func (r *LocalFileResolver) ResolveDiscovery(domain string) (*types.DiscoveryDocument, error) {
	path := filepath.Join(r.DiscoveryDir, domain+".json")
	data, err := os.ReadFile(path)
	if err != nil {
		return nil, fmt.Errorf("cannot read %s: %w", path, err)
	}
	var doc types.DiscoveryDocument
	if err := json.Unmarshal(data, &doc); err != nil {
		return nil, err
	}
	return &doc, nil
}

// ResolveRevocation implements DiscoveryResolver.
func (r *LocalFileResolver) ResolveRevocation(domain string, _ *types.DiscoveryDocument) (*types.RevocationDocument, error) {
	dir := r.RevocationDir
	if dir == "" {
		dir = r.DiscoveryDir
	}
	path := filepath.Join(dir, domain+".revocations.json")
	data, err := os.ReadFile(path)
	if err != nil {
		if errors.Is(err, os.ErrNotExist) {
			return nil, nil
		}
		return nil, fmt.Errorf("cannot read %s: %w", path, err)
	}
	var doc types.RevocationDocument
	if err := json.Unmarshal(data, &doc); err != nil {
		return nil, err
	}
	return &doc, nil
}

// TrustBundleResolver resolves documents from a pre-loaded TrustBundle.
type TrustBundleResolver struct {
	discovery   map[string]*types.DiscoveryDocument
	revocations map[string]*types.RevocationDocument
}

// NewTrustBundleResolver indexes b by entity for O(1) lookup.
func NewTrustBundleResolver(b *types.TrustBundle) *TrustBundleResolver {
	r := &TrustBundleResolver{
		discovery:   make(map[string]*types.DiscoveryDocument, len(b.Documents)),
		revocations: make(map[string]*types.RevocationDocument, len(b.Revocations)),
	}
	for i := range b.Documents {
		d := b.Documents[i]
		r.discovery[d.Entity] = &d
	}
	for i := range b.Revocations {
		rev := b.Revocations[i]
		r.revocations[rev.Entity] = &rev
	}
	return r
}

// TrustBundleResolverFromJSON builds a TrustBundleResolver from a JSON-encoded
// TrustBundle.
func TrustBundleResolverFromJSON(data []byte) (*TrustBundleResolver, error) {
	var b types.TrustBundle
	if err := json.Unmarshal(data, &b); err != nil {
		return nil, err
	}
	return NewTrustBundleResolver(&b), nil
}

// ResolveDiscovery implements DiscoveryResolver.
func (r *TrustBundleResolver) ResolveDiscovery(domain string) (*types.DiscoveryDocument, error) {
	d, ok := r.discovery[domain]
	if !ok {
		return nil, fmt.Errorf("domain '%s' not in trust bundle", domain)
	}
	cp := *d
	return &cp, nil
}

// ResolveRevocation implements DiscoveryResolver.
func (r *TrustBundleResolver) ResolveRevocation(domain string, _ *types.DiscoveryDocument) (*types.RevocationDocument, error) {
	rev, ok := r.revocations[domain]
	if !ok {
		return nil, nil
	}
	cp := *rev
	return &cp, nil
}

// ChainResolver tries a sequence of resolvers in order until one succeeds.
type ChainResolver struct {
	Resolvers []DiscoveryResolver
}

// NewChainResolver builds a ChainResolver from a slice of resolvers.
func NewChainResolver(resolvers []DiscoveryResolver) *ChainResolver {
	return &ChainResolver{Resolvers: resolvers}
}

// ResolveDiscovery implements DiscoveryResolver.
func (r *ChainResolver) ResolveDiscovery(domain string) (*types.DiscoveryDocument, error) {
	var lastErr error = errors.New("no resolvers configured")
	for _, sub := range r.Resolvers {
		doc, err := sub.ResolveDiscovery(domain)
		if err == nil {
			return doc, nil
		}
		lastErr = err
	}
	return nil, lastErr
}

// ResolveRevocation implements DiscoveryResolver.
func (r *ChainResolver) ResolveRevocation(domain string, doc *types.DiscoveryDocument) (*types.RevocationDocument, error) {
	for _, sub := range r.Resolvers {
		rev, err := sub.ResolveRevocation(domain, doc)
		if err != nil {
			continue
		}
		if rev != nil {
			return rev, nil
		}
	}
	return nil, nil
}
